proxy/coco/src/peer.rs

```diff
@@ -123,385 +122,441 @@ impl Peer {
+    request_sender: mpsc::Sender<RequestEvent>,
+
+    announcements: mpsc::Receiver<AnnounceEvent>,
+    peer_syncs: mpsc::Receiver<SyncEvent>,
+    timeouts: mpsc::Receiver<TimeoutEvent>,
+    requests: mpsc::Receiver<RequestEvent>,
+}
+
+impl Subroutines {
+    /// Constructor.
+    pub fn new(
+        state: State,
+        store: kv::Store,
+        waiting_room: Shared<WaitingRoom<Instant, Duration>>,
+        run_config: RunConfig,
+        protocol_events: BoxStream<'static, ProtocolEvent<Gossip>>,
+        subscriber: broadcast::Sender<Event>,
+    ) -> Self {
+        let announce_timer = interval(run_config.announce.interval);
+        let ping_timer = interval(PING_PERIOD);
+
+        let (announce_sender, announcements) = mpsc::channel::<AnnounceEvent>(RECEIVER_CAPACITY);
+        let (peer_sync_sender, peer_syncs) = mpsc::channel::<SyncEvent>(RECEIVER_CAPACITY);
+        let (timeout_sender, timeouts) = mpsc::channel::<TimeoutEvent>(RECEIVER_CAPACITY);
+        let (request_sender, requests) = mpsc::channel::<RequestEvent>(RECEIVER_CAPACITY);
+
+        let run_state = RunState::from(run_config);
+
+        let request_queries = waiting_room::stream::Queries::new(waiting_room.clone().value);
+        let request_clones = waiting_room::stream::Clones::new(waiting_room.clone().value);
+
+        Self {
+            tasks: FuturesUnordered::new(),
+
+            state,
+            store,
+            waiting_room,
+            run_state,
+
+            protocol_events: protocol_events.boxed(),
+
+            announce_timer,
+            ping_timer,
+
+            subscriber,
+
+            request_queries,
+            request_clones,
+
+            announce_sender,
+            peer_sync_sender,
+            timeout_sender,
+            request_sender,
+
+            announcements,
+            peer_syncs,
+            timeouts,
+            requests,
         }
     }
+}
 
-    Ok(())
+impl Drop for Subroutines {
+    fn drop(&mut self) {
+        for task in self.tasks.iter_mut() {
+            task.abort()
+        }
+    }
 }
 
-/// Announcement subroutine.
-fn announce(
-    state: State,
-    store: kv::Store,
-    mut sender: mpsc::Sender<AnnounceEvent>,
-    stop: impl Future<Output = ()> + Send + Unpin + 'static,
-) {
-    tokio::spawn(async move {
-        let go = async {
-            match announcement::run(&state, &store).await {
-                Ok(updates) => {
-                    sender.send(AnnounceEvent::Succeeded(updates)).await.ok();
-                },
-                Err(err) => {
-                    log::error!("announce error: {:?}", err);
-                    sender.send(AnnounceEvent::Failed).await.ok();
-                },
+impl Future for Subroutines {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut events = Vec::with_capacity(9);
+
+        // FIXME(kim): can we not somehow coalesce all of these into a stream?
+        {
```

Comment:
This makes it easier to reason about than the `select!` stuff :ok_hand:

Reply:
Ya, although it is presumably less efficient. The expansion of `tokio::select!` is incomprehensible, but the `futures` equivalent seems to do something more smarterer than polling all the futures in sequence.

I would prefer to instead have the "subroutine" tasks return a value, so we can just poll the `FuturesUnordered`. This would allow the get rid of this bouquet of channels. The only reason this doesn't work at the moment is that `sync` task wants to dispatch a message that it's about to start -- but perhaps this could be restructured somehow?

Reply:
Another possibility would be to communicate via a single channel of type `Event`. Some type information would get lost, but I guess that'd be fine, as it's just `Subroutines` communicating with itself.

Reply:
> This makes it easier to reason about than the `select!` stuff :ok_hand:

Find it hard to agree with that, but than again I learned systems programming from Mr Pike.

> I would prefer to instead have the "subroutine" tasks return a value, so we can just poll the `FuturesUnordered`. This would allow the get rid of this bouquet of channels. The only reason this doesn't work at the moment is that `sync` task wants to dispatch a message that it's about to start -- but perhaps this could be restructured somehow?

If we rework the subroutines to produce values, we would assume to have new input `Event`s after completing the commands produced by `transition`. Wouldn't be possible to stick `SyncEvent::Started` into that input-output loop before starting the sync routine?

Reply:
> Wouldn't be possible to stick SyncEvent::Started into that input-output loop before starting the sync routine?

I'm not sure I understand that, but then, I'm also a bit iffy on the semantics of the `RunState`-machine: calls to `transition` can happen in any order, depending on when tasks finish. In that sense, every "subroutine" is treated as being independent -- except that we're `select`ing over completion events, so they all get coalesced into a single stream anyway. The difference to using an _actual_ single channel would be that the order in which the events come out the other end depends on task completion, whereas the giant select introduces bias as to which channel gets polled first.

Reply:
> calls to `transition` can happen in any order, depending on when tasks finish

Exactly, I didn't mean to oppose the single channel idea. I was commenting on your point that the only thing that stands in the way of getting rid of the channels is the fact that we want to send the `SyncEvent::Started` in the sync routine - where I think we can do that outside of the routine.

> whereas the giant select introduces bias as to which channel gets polled first.

What's the bias on the select? Both `futures::select` as well as `tokio::select` have some randomness for branch selection:

> If multiple futures are ready, one will be pseudo-randomly selected at runtime.

> select! randomly picks a branch to check first. This provides some level of fairness when calling select! in a loop with branches that are always ready.

Reply:
@xla The bias is obvious with this change, which polls the channels in a fixed order. With `select!`, you get a random result out of all ready channels -- as you cite, this prevents the first in the list to starve all others, but that is not the same as completion order.

Anyways, I experimented with coalescing all events into a `SelectAll` (which supposedly has the same semantics as `select!`), but that somehow causes the waiting room streams to not be woken (regardless of your `Ping` userspace waker). That's a bit puzzling, but I guess that waiting room stuff needs a tad bit more async awareness -- we actually know when to wake, which is whenever something is inserted. For now it means that we can't go down that route, however.

Reply:
Here's a [demo](https://github.com/radicle-dev/radicle-upstream/compare/kim/vp-of-tasks...souproutine-coalesce)

Reply:
> @xla The bias is obvious with this change, which polls the channels in a fixed order. With `select!`, you get a random result out of all ready channels -- as you cite, this prevents the first in the list to starve all others, but that is not the same as completion order.

I follow, as you pointed out the ordering of inputs (completion order) is not significant for the RunState transitions, so it shouldn't have a noticable effect.

> Anyways, I experimented with coalescing all events into a `SelectAll` (which supposedly has the same semantics as `select!`), but that somehow causes the waiting room streams to not be woken (regardless of your `Ping` userspace waker). That's a bit puzzling, but I guess that waiting room stuff needs a tad bit more async awareness -- we actually know when to wake, which is whenever something is inserted. For now it means that we can't go down that route, however.

I'm looking into #981 at the moment and currently am tempted to drop the `Stream` impls entirely. Instead just have intervales like we have for the announcements and wait for those to complete. Upon completion inspect the waiting room. This would get rid of this whole unlocking inside of poll and waker business. Unfortunately the semantics of when to wake is not as straightforward as insertion - there is also the case of rescheduling which does not depend on a mutation of the waiting room. For that case it needs to know if the time since the last query has exceeded a certain duration.

Would the removal of the `Stream` impls for the two waiting room concerns enable moving to `SelectAll`?

Reply:
> Instead just have intervales like we have for the announcements and wait for those to complete

Ya, if that works it's the simpler model.

> Unfortunately the semantics of when to wake is not as straightforward as insertion

I was looking at the internal `requests` state -- whenever something is inserted there, readers could be woken. But yeah, that requires some notification machinery, so `WaitingRoom` needs to be async-aware. Iiuc, `WaitingRoom` is quite similar to `RunState`, in that it decides on tasks to be scheduled, while itself being just a pure state machine. Perhaps the `RunState` could just have a `WaitingRoom` internally, and advance its state on `transition()`?

Reply:
> I was looking at the internal `requests` state -- whenever something is inserted there, readers could be woken. But yeah, that requires some notification machinery, so `WaitingRoom` needs to be async-aware. Iiuc, `WaitingRoom` is quite similar to `RunState`, in that it decides on tasks to be scheduled, while itself being just a pure state machine. Perhaps the `RunState` could just have a `WaitingRoom` internally, and advance its state on `transition()`?

It might be the only sensible way forward, this locking business is pooping at all ze parties. The reason access to the waiting room is not serialised is that we also hand it to the API so user action can trigger the creation of requests. We could get rid of that by handing out something that pos a new message onto a channel which we use as another input the RunState state machine. Let me conjure something up to see if it twers. :mage_man: